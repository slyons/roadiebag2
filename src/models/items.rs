use chrono::Utc;
use sea_orm::entity::prelude::*;
pub use super::_entities::items::{self, Entity, ActiveModel, Model};
use loco_rs:: {
    model::{ModelError, ModelResult},
    validation,
    validator::Validate,
};
use sea_orm::{entity::prelude::*, ActiveValue, DatabaseConnection, DbErr, TransactionTrait, QuerySelect, Paginator, QueryOrder};
use sea_orm::sea_query::{*, extension::postgres::PgExpr};
use serde::{Deserialize, Serialize};

#[derive(Debug, Validate, Deserialize)]
pub struct ModelValidator {
    #[validate(length(min=1, message="Name must be at least 1 character long"))]
    pub name: String,
    #[validate(range(min=1))]
    pub quantity: i32,
    #[validate(range(min=0, max=3))]
    pub size: i16
}

impl From<&ActiveModel> for ModelValidator {
    fn from(value: &ActiveModel) -> Self {
        Self {
            name: value.name.as_ref().to_string(),
            quantity: *value.quantity.as_ref(),
            size: *value.size.as_ref()
        }
    }
}

impl Into<interface::Item> for Model {
    fn into(self) -> interface::Item {
        interface::Item {
            created_at: self.created_at,
            updated_at: self.updated_at,
            name: self.name,
            description: self.description,
            id: self.id,
            size: interface::ItemSize::from_repr(self.size).unwrap(),
            infinite: self.infinite,
            quantity: self.quantity
        }
    }
}

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    // extend activemodel below (keep comment for generators)
    async fn before_save<C>(self, _db: &C, insert: bool) -> Result<Self, DbErr>
        where C: ConnectionTrait
    {
        {
            self.validate()?;
            let mut this = self;

            if insert {
                this.created_at = ActiveValue::Set(Utc::now().naive_utc());
            }
            this.updated_at = ActiveValue::Set(Utc::now().naive_utc());
            Ok(this)
        }
    }
}

impl Model {
    pub async fn find_by_id(db: &DatabaseConnection, id: i32) -> ModelResult<Self> {
        let item = items::Entity::find()
            .filter(items::Column::Id.eq(id))
            .one(db)
            .await?;
        item.ok_or_else(|| ModelError::EntityNotFound)
    }

    pub async fn create(db: &DatabaseConnection, create: interface::CreateUpdateItem) -> ModelResult<Self> {
        let txn = db.begin().await?;
        let item = items::ActiveModel {
            name: ActiveValue::Set(create.name),
            description: ActiveValue::Set(create.description),
            quantity: ActiveValue::Set(create.quantity),
            size: ActiveValue::Set(create.size as i16),
            infinite: ActiveValue::Set(create.infinite),
            ..Default::default()
        }
            .insert(&txn)
            .await?;

        txn.commit().await?;
        Ok(item)
    }

    pub async fn update(db: &DatabaseConnection, id: i32, update: interface::CreateUpdateItem) -> ModelResult<Self> {
        let txn = db.begin().await?;

        if items::Entity::find()
            .filter(items::Column::Id.eq(id))
            .one(&txn)
            .await?
            .is_none()
        {
            return Err(ModelError::EntityNotFound {});
        }

        let item = items::ActiveModel {
            name: ActiveValue::Set(update.name),
            description: ActiveValue::Set(update.description),
            quantity: ActiveValue::Set(update.quantity),
            size: ActiveValue::Set(update.size as i16),
            infinite: ActiveValue::Set(update.infinite),
            id: ActiveValue::Set(id),
            ..Default::default()
        }
            .update(&txn)
            .await?;
        txn.commit().await?;
        Ok(item)
    }

    pub async fn delete(db: &DatabaseConnection, id: i32) -> ModelResult<()> {
        let txn = db.begin().await?;

        if items::Entity::find()
            .filter(items::Column::Id.eq(id))
            .one(&txn)
            .await?
            .is_none()
        {
            return Err(ModelError::EntityNotFound {});
        }

        let item = items::ActiveModel {
            id: ActiveValue::Set(id),
            ..Default::default()
        }
            .delete(&txn)
            .await?;

        txn.commit().await?;
        Ok(())
    }

    pub async fn list(db: &DatabaseConnection, filter: Option<interface::ItemFilter>) -> ModelResult<interface::ItemPage> {
        let filter = filter.unwrap_or_default();

        let mut query = items::Entity::find()
            .order_by_desc(items::Column::Id);
        if let Some(mut name) = filter.name {
            if !name.contains("%") {
                name = name + "%";
            }
            query = query.filter(Expr::col(items::Column::Name).ilike(name));
        }

        if let Some(mut description) = filter.description {
            if !description.contains("%") {
                description = description + "%";
            }
            query = query.filter(Expr::col(items::Column::Description).ilike(description));
        }

        if let Some(size) = filter.size {
            let item_size = size as i16;
            query = query.filter(items::Column::Size.eq(item_size));
        }

        if let Some(infinite) = filter.infinite {
            query = query.filter(items::Column::Infinite.eq(infinite));
        }

        let page_size = if filter.page_size > 0 {
            filter.page_size
        } else {
            50
        };

        let paginator = query.paginate(db, page_size);
        let items_and_pages = paginator.num_items_and_pages().await?;
        let items = paginator
            .fetch_page(filter.page_num)
            .await?
            .into_iter()
            .map(Into::into)
            .collect();

        Ok(interface::ItemPage {
            items: items,
            page_num: filter.page_num,
            page_size: page_size,
            total_pages: items_and_pages.number_of_pages,
            total_results: items_and_pages.number_of_items
        })
    }
}

impl ActiveModel {
    pub fn validate(&self) -> Result<(), DbErr> {
        let validator: ModelValidator = self.into();
        validator
            .validate()
            .map_err(|e| validation::into_db_error(&e))
    }
}