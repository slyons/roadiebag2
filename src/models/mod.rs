pub mod _entities;
//pub mod notes;
pub mod users;

pub mod items;
pub mod taken_items;
