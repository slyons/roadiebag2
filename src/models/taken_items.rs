use chrono::Utc;
use interface::TakenItem;
use loco_rs::model::{ModelError, ModelResult};
use sea_orm::{
    entity::prelude::*,
    sea_query::{Alias, Query},
    ActiveValue, JoinType, QuerySelect, TransactionTrait,
};

use super::_entities::items;
pub use super::_entities::taken_items::{self, ActiveModel, Entity, Model};

#[async_trait::async_trait]
impl ActiveModelBehavior for ActiveModel {
    async fn before_save<C>(self, _db: &C, insert: bool) -> Result<Self, DbErr>
    where
        C: ConnectionTrait,
    {
        {
            let mut this = self;

            if insert {
                this.created_at = ActiveValue::Set(Utc::now().naive_utc());
            }
            this.updated_at = ActiveValue::Set(Utc::now().naive_utc());
            Ok(this)
        }
    }
}

impl Into<interface::TakenItem> for Model {
    fn into(self) -> TakenItem {
        TakenItem {
            created_at: self.created_at,
            updated_at: self.updated_at,
            id: self.id,
            item_id: self.item_id,
            rounds_left: self.rounds_left,
            done: self.done,
            rounds_total: self.rounds_total,
        }
    }
}

impl Model {
    pub async fn get_current(db: &DatabaseConnection) -> ModelResult<Option<interface::TakenItem>> {
        Ok(taken_items::Entity::find()
            .filter(taken_items::Column::Done.eq(false))
            .one(db)
            .await?
            .map(Into::into))
    }

    pub async fn decrement_rounds(
        db: &DatabaseConnection,
    ) -> ModelResult<Option<interface::TakenItem>> {
        let current_item = Model::get_current(db).await?;
        if let Some(itm) = current_item {
            let new_round_count = itm.rounds_left - 1;
            let done = new_round_count <= 0;
            let model = ActiveModel {
                id: ActiveValue::Set(itm.id),
                rounds_left: ActiveValue::Set(new_round_count),
                done: ActiveValue::Set(done),
                ..Default::default()
            }
            .update(db)
            .await?;
            Ok(Some(model.into()))
        } else {
            Ok(current_item)
        }
    }

    pub async fn mark_done(db: &DatabaseConnection) -> ModelResult<()> {
        let current_item = Model::get_current(db).await?;
        if let Some(itm) = current_item {
            let model = ActiveModel {
                id: ActiveValue::Set(itm.id),
                done: ActiveValue::Set(true),
                ..Default::default()
            }
            .update(db)
            .await?;
            Ok(())
        } else {
            Ok(())
        }
    }

    pub async fn get_random(db: &DatabaseConnection) -> ModelResult<interface::TakenItem> {
        let existing = Model::get_current(db).await?;
        if let Some(ext) = existing {
            tracing::info!("Current item is {:?}", ext);
            Ok(ext)
        } else {
            let txn = db.begin().await?;
            let item_uses = items::Entity::find()
                .left_join(taken_items::Entity)
                .group_by(items::Column::Id)
                .to_owned()
                .having(
                    Expr::expr(
                        Expr::case(
                            items::Column::Infinite.eq(false),
                            items::Column::Quantity.into_expr().sub(
                                Expr::expr(taken_items::Column::Id.into_expr().count()).if_null(0),
                            ),
                        )
                        .finally(1),
                    )
                    .gte(1),
                );
            //.column(items::Column::Id);
            let item_count = item_uses.clone().count(&txn).await?;
            tracing::info!("Item count is {}", item_count);
            let item_offset = rand::random::<u64>().clamp(0, item_count - 1);
            tracing::info!("Selecting offset {}", item_offset);
            //let item_id = item_uses.offset(item_offset).into_tuple().one(&txn)
            //    .await?
            //    .expect(&format!("Item with offset {} returned None", item_offset));
            let item = item_uses
                .offset(item_offset)
                .one(&txn)
                .await?
                .expect(&format!("Item with offset {} returned None", item_offset));
            let total_rounds = rand::random::<i16>().clamp(1, 6);
            let model = ActiveModel {
                item_id: ActiveValue::Set(item.id),
                rounds_left: ActiveValue::Set(total_rounds),
                rounds_total: ActiveValue::Set(total_rounds),
                done: ActiveValue::Set(false),
                ..Default::default()
            }
            .insert(&txn)
            .await?;

            txn.commit().await?;
            Ok(model.into())
        }
    }
}
