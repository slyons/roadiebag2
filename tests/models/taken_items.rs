use insta::assert_debug_snapshot;
use roadiebag2::app::App;
use loco_rs::testing;
use serial_test::serial;
use test_log::test;
use tracing_test::traced_test;
use roadiebag2::models::{items, taken_items};

macro_rules! configure_insta {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_prepend_module_to_snapshot(false);
        let _guard = settings.bind_to_scope();
    };
}

#[tokio::test]
#[serial]
#[traced_test]
async fn test_model() {
    configure_insta!();

    let boot = testing::boot_test::<App>().await.unwrap();
    testing::seed::<App>(&boot.app_context.db).await.unwrap();

    let create = interface::CreateUpdateItem {
        name: "Test item".to_string(),
        description: None,
        quantity: 2,
        size: interface::ItemSize::Small,
        infinite: false
    };

    let model = items::Model::create(&boot.app_context.db, create).await;

    let current_taken_none = taken_items::Model::get_current(&boot.app_context.db).await.unwrap();

    let current_random = taken_items::Model::get_random(&boot.app_context.db).await.unwrap();

    taken_items::Model::mark_done(&boot.app_context.db).await.unwrap();

    let current_taken_done = taken_items::Model::get_current(&boot.app_context.db).await.unwrap();


    insta::with_settings!({
        filters => {
            let mut fvec = testing::CLEANUP_DATE.to_vec();
            fvec.extend(vec![
                (r"rounds_left: \d", "rounds_left: ROUNDS_LEFT"),
                (r"rounds_total: \d", "rounds_total: ROUNDS_TOTAL")
            ]);
            fvec
        }
    }, {
        assert_debug_snapshot!((
            current_taken_none,
            current_random,
            current_taken_done
        ))
    });
}

#[tokio::test]
#[serial]
async fn test_decrement() {
    configure_insta!();

    let boot = testing::boot_test::<App>().await.unwrap();
    testing::seed::<App>(&boot.app_context.db).await.unwrap();

    let create = interface::CreateUpdateItem {
        name: "Test item".to_string(),
        description: None,
        quantity: 2,
        size: interface::ItemSize::Small,
        infinite: false
    };

    let _model = items::Model::create(&boot.app_context.db, create).await;
    let current_random = taken_items::Model::get_random(&boot.app_context.db).await.unwrap();

    let decr = taken_items::Model::decrement_rounds(&boot.app_context.db).await.unwrap().unwrap();

    assert_eq!(current_random.rounds_left - 1, decr.rounds_left)
}

