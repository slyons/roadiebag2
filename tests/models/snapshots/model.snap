---
source: tests/models/taken_items.rs
expression: "(current_taken_none, current_random, current_taken_done)"
---
(
    None,
    TakenItem {
        created_at: DATE,
        updated_at: DATE,
        id: 1,
        item_id: 1,
        rounds_left: ROUNDS_LEFT,
        done: false,
        rounds_total: ROUNDS_TOTAL,
    },
    None,
)
